text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+- Feature Name: `cargo-script`
+- Start Date: 2023-03-31
+- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
+- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
+- RFC PR: [rust-lang/rfcs#3502](https://github.com/rust-lang/rfcs/pull/3502)
+- Rust Issue: [rust-lang/cargo#12207](https://github.com/rust-lang/cargo/issues/12207)
+
+# Summary
+[summary]: #summary
+
+This RFC adds support for single-file bin packages in cargo.
+Single-file bin packages are rust source files with an embedded manifest and a
+`main`.
+These files will be accepted by cargo commands as `--manifest-path` just like `Cargo.toml` files.
+`cargo` will be modified to accept `cargo <file>.rs` as a shortcut to `cargo
+run --manifest-path <file>.rs`;
+this allows placing `cargo` in a `#!` line for directly running these files.
```

Comment:
Just realized a potentially fatal issue with using cargo like this: what if the program doesn't have an extension, and it's named `test` (not uncommon), or `bench`, or for that matter *any future name we want to use`.

Reply:
This is covered in the `cargo <file>.rs` section where we talk about precedence and disambiguation.

For extensionless scripts, you must have a `/` in the parameter, so `cargo ./test`.  From what I understand of shebangs being run, this will always be the case.

Reply:
We may need a dedicated script runner program for this reason.

Reply:
Which cases are you concerned about?

Reply:
On some platforms it is possible for the script to be passed as just "basename" with no ./ or other path. (I'll look further into when and how that can happen.)

Reply:
I tested to confirm: On Linux, running a script in the current directory with `PATH` unset or set to the empty string will result in passing the basename to the interpreter with no `./` prepended.

There may be other OSes and situations in which this behavior applies.

I'm not sure if any of these scenarios are important enough to warrant a different entry point, but I think it's worth explicitly documenting this issue and confirming that we can live with it.

Reply:
btw I looked into how other tools handle ambiguity: https://github.com/epage/cargo-script-mvs/issues/154

dlang's dub was the easier to decipher based on the source

Reply:
I assume whats happening here is the CWD is being considered part of  PATH, like on Windows.  I glossed over this because my understanding is its generally frowned upon.  Its now called out in f7b8bf3e along with a couple other potential problems.

Reply:
> On some platforms it is possible for the script to be passed as just "basename" with no ./ or other path. (I'll look further into when and how that can happen.)

> On Linux, running a script in the current directory with PATH unset or set to the empty string will result in passing the basename to the interpreter with no ./ prepended.

This is actually down to the shell on unix - see e.g. https://github.com/fish-shell/fish-shell/issues/9143.

Basically whatever ends up being passed to some `execXY` or other.

Reply:
That seems to be about absolute vs relative which I think is a reasonable thing for shells to choose one way or the other.  The proposal has no problem with that.

The problem here is when dealing with `foo` without any `/` in it.  Shells and the kernel are limited on this.  If CWD isn't in the path, they have to prepend the path for it to be able to be run.  If CWD is in the path, well, then its a problem.

Reply:
I believe it is possible for it to not have a `/`, even if $PATH is not empty. If you'll forgive me, my demonstration is C (and I'm not actually any good at C):

In test.c, compile with `gcc test.c -o test`:

```C
#include <unistd.h>
#include <stddef.h>
  
int main(int argc, char **argv) {
  char *const args[2] = {"HAHAHA", NULL};
  char **env;
  extern char **environ;
 
  execve("foo", args, env);
}
```

"nameprinter.c", compile as `gcc nameprinter.c -o nameprinter`:

```C
#include <stdio.h>
  
int main(int argc, char **argv) {
    for (int i = 0; i < argc; i++) {
        printf("%d: %s\n", i, argv[i]);
    }
    return 0;
}
```

And a file called "foo", where the contents are

```
#!/absolute/path/to/nameprinter
```

(alas you'll have to add that yourself)

Make everything executable and run `PATH=/bin ./test`.

This gets me output of:

```
0: /home/alfa/nameprinter
1: foo
```

So nameprinter receives the argument "foo", without any `/`.

Up-to-date Archlinux machine, kernel 6.5.5, glibc 2.38.

Reply:
Let me check my understanding of the situation.  Someone can run into this if they invoke a script via a  `execve` because it isn't looking things up by `PATH` but is being given a path to the binary (so `./` can be assumed).

Reply:
Yes, and the execve can be in anything. It could be in a shell, it could be in another program that runs the script for some reason (tho not via `system(3)` or equivalent because that would go via /bin/sh, which would need `./` - unless that /bin/sh removes the `./` for some reason), it could be in another interpreter.

The point being that there's no guarantee for the `/`.

You would use a valid `execve()` invocation to run the file, and it could end up running a cargo builtin.

Reply:
This is now at least noted as of ad5ab232

Reply:
We talked about this more in the cargo team meeting.  Overall, we feel that the chance of conflict will be very small and can be worked around, so we don't see anything immediately done about it.

We did talk a bit [AT_EXECFN](https://www.man7.org/linux/man-pages/man3/getauxval.3.html) and have put that down as an [Unresolved questions](https://github.com/epage/rfcs/blob/script/text/3502-cargo-script.md#unresolved-questions) to be worked out during stabilization for whether we should leverage it for detecting that we were run via shebang